File: QuickStart/QuickStart/AccountViewController.m
/* Copyright (c) 2018 BlackBerry.  All Rights Reserved.
* 
* Licensed under the Apache License, Version 2.0 (the "License"); 
* you may not use this file except in compliance with the License. 
* You may obtain a copy of the License at 
* 
* http://www.apache.org/licenses/LICENSE-2.0 
* 
* Unless required by applicable law or agreed to in writing, software 
* distributed under the License is distributed on an "AS IS" BASIS, 
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and 
* limitations under the License. 
  
* This sample code was created by BlackBerry using SDKs from Apple Inc. 
* and may contain code licensed for use only with Apple products. 
* Please review your Apple SDK Agreement for additional details. 
*/ 

#import "AccountViewController.h"
#import "BBMConfigManager.h"

#import "BBMAccess.h"
#import "BBMAuthController.h"
#import "BBMAuthenticatedAccount.h"
#import "BBMEndpointManager.h"
#import "BBMAppUserListener.h"
#import "BBMUserManager.h"
#import "UIView+Extra.h"

#import <BBMEnterprise/BBMEnterprise.h>

#import "BBMGoogleTokenManager.h"
#import <GoogleSignIn/GoogleSignIn.h>
#import <GoogleSignIn/GIDSignInButton.h>
#import "BBMAzureTokenManager.h"
#import "BBMAzureUserManager.h"

#import "BBMTestTokenManager.h"

@interface AccountViewController () <BBMConnectivityListener, BBMAuthControllerDelegate>

@property (weak, nonatomic) IBOutlet UILabel *serviceStateLabel;
@property (weak, nonatomic) IBOutlet UILabel *authTokenStateLabel;
@property (weak, nonatomic) IBOutlet UILabel *setupStateLabel;
@property (weak, nonatomic) IBOutlet UILabel *regIdLabel;
@property (weak, nonatomic) IBOutlet UILabel *domainLabel;
@property (weak, nonatomic) IBOutlet UILabel *userEmailLabel;
@property (weak, nonatomic) IBOutlet UILabel *serviceConnectivityLabel;

@property (weak, nonatomic) IBOutlet UIView   *signInButtonContainer;
@property (weak, nonatomic) IBOutlet UIButton *signOutButton;

@property (nonatomic, strong) GIDSignInButton *googleSignInButton;
@property (nonatomic, strong) UIButton        *signInButton;

@property (nonatomic, strong) BBMAuthController *authController;
@property (nonatomic, strong) ObservableMonitor *serviceMonitor;

@end

@implementation AccountViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    Class tokenManager;
    if([BBMConfigManager defaultManager].type == kGoogleSignIn)
    {
        self.googleSignInButton = [[GIDSignInButton alloc] initWithFrame:self.signInButtonContainer.bounds];
        [self.signInButtonContainer addSubviewAndContraintsWithSameFrame:self.googleSignInButton];
        tokenManager = [BBMGoogleTokenManager class];
    }
    else if([BBMConfigManager defaultManager].type == kAzureAD)
    {
        self.signInButton = [[UIButton alloc] init];
        [self.signInButton setTitle:@"Azure AD Sign In" forState:UIControlStateNormal];
        [self.signInButtonContainer addSubviewAndContraintsWithSameFrame:self.signInButton];
        [self.signInButton addTarget:self
                                   action:@selector(signIn:)
                         forControlEvents:UIControlEventTouchUpInside];
        [self.view layoutIfNeeded];
        self.signInButtonContainer.backgroundColor = [UIColor blueColor];
        tokenManager = [BBMAzureTokenManager class];
    }
    else if([BBMConfigManager defaultManager].type == kTestAuth)
    {
        self.signInButton = [[UIButton alloc] init];
        [self.signInButton setTitle:@"Sign In" forState:UIControlStateNormal];
        [self.signInButtonContainer addSubviewAndContraintsWithSameFrame:self.signInButton];
        [self.signInButton addTarget:self
                              action:@selector(signIn:)
                    forControlEvents:UIControlEventTouchUpInside];
        [self.view layoutIfNeeded];
        self.signInButtonContainer.backgroundColor = [UIColor blueColor];
        tokenManager = [BBMTestTokenManager class];
    }else{
        NSAssert(NO, @"Unsupported identity provider");
    }


    //Configure our buttons and labels
    self.signInButtonContainer.hidden = YES;
    self.signOutButton.hidden = YES;
    self.domainLabel.text = [BBMConfigManager defaultManager].sdkServiceDomain;

    //Listen for connectivity changes to the BBM Enterprise Service
    [[BBMEnterpriseService service] addConnectivityListener:self];



    //Create an AuthController instance.  This view controller will act as the delegate and the
    //root controller for any modally presented auth UI.  We will use the BBMGoogleTokenManager
    //(a wrapper around the GoolgeSignIn API) to fetch tokens
    self.authController = [[BBMAuthController alloc] initWithTokenManager:tokenManager
                                                               userSource:nil
                                                       keyStorageProvider:nil
                                                                   domain:[BBMConfigManager defaultManager].sdkServiceDomain
                                                              environment:[BBMConfigManager defaultManager].environment];

    self.authController.rootController = self;

    //Auth controller using delegation
    [self.authController addDelegate:self];


    /***********

    //Alternative: Auth controller using ObservableMonitor.  This approach monitors the serviceStarted
    //and authState properties on the authManager instance rather than registering the view controller
    //as an explicit delegate

    typeof(self) __weak weakSelf = self;
    self.serviceMonitor = [ObservableMonitor monitorActivatedWithName:@"serviceMonitor" block:^{
        [weakSelf serviceStateChanged:weakSelf.authController.serviceStarted];
        [weakSelf authStateChanged:weakSelf.authController.authState];
    }];
    
    *************/


    //Start the BBM Enterprise service.
    [self.authController startBBMEnterpriseService];

    //Resume our previous session
    [self.authController signInSilently];
}


#pragma mark - BBMAuthDelegate

//The AuthController will invoke this callback any time the user credentials/auth state changes
- (void)authStateChanged:(BBMAuthState *)authState
{
    self.authTokenStateLabel.text = authState.authTokenState ?: @"No Token";
    self.setupStateLabel.text = authState.setupState ?: @"Setup Not Started";
    self.regIdLabel.text = authState.regId.integerValue ? authState.regId.stringValue : @"";
    self.userEmailLabel.text = authState.account.email;

    self.signInButtonContainer.hidden = self.authController.startedAndAuthenticated;
    self.signOutButton.hidden = !self.authController.startedAndAuthenticated;

    //In order to complete the setup process, we must synchronize the profile keys.  For the sake of
    //brevity, we will simply set the key state to synced which allows setup to complete without setting up
    //a complete key provider.  See the other BBM Enterprise samples and/or BBMKeyManager in
    // /examples/support for details on how to export/import profile keys.
    //This is not required if you're using the default Spark Communications Key Management Service in which case, the
    //keys will be synced automatically after entering your application password
    if(![BBMConfigManager defaultManager].isUsingBlackBerryKMS &&
       authState.setupState == kBBMSetupStateOngoing &&
       [[[BBMEnterpriseService service] model] globalProfileKeysState] == kBBMKeyStateNotSynced)
    {
        NSDictionary* element = @{@"name":@"profileKeysState", @"value": @"Synced" };
        BBMRequestListChangeMessage *syncProfileKey = [[BBMRequestListChangeMessage alloc] initWithElements:@[element] type:@"global"];
        [[BBMEnterpriseService service] sendMessageToService:syncProfileKey];
    }

    //Our token has been rejected
    if(authState.authTokenState == kBBMAuthStateRejected) {
        self.signOutButton.hidden = NO;
        self.signInButtonContainer.hidden = YES;
    }

    //We're logged in but setup of BBM hasn't completed yet
    if(authState.setupState == kBBMSetupStateOngoing) {
        self.signInButtonContainer.hidden = YES;
        self.signOutButton.hidden = NO;
    }

    // If state is full, ask for list of endpoints and remove one so that setup can continue.
    // In practice, the user should be prompted to choose an endpoint to log out.
    if(authState.setupState == kBBMSetupStateFull) {
        [self.authController.endpointManager deregisterAnyEndpointAndContinueSetup];
    }

    [self.tableView reloadData];
}

- (void)serviceStateChanged:(BOOL)serviceStarted
{
    self.serviceStateLabel.text = serviceStarted ? @"Started" : @"Stopped";
    [self.tableView reloadData];
}

#pragma mark - BBMConnectivityListener

- (void)connectivityStateChange:(BOOL)connected strict:(BOOL)connectedStrict
{
    self.serviceConnectivityLabel.text = connected ? @"Connected" : @"Disconnected";
}


#pragma mark - UI Actions


- (IBAction)signOut:(id)sender
{
    //Wipe all local BBM data
    [[BBMEnterpriseService service] resetService];
    [self.authController signOut];
}

- (IBAction)signIn:(id)sender
{
    [self.authController.tokenManager signIn];
}



@end

File: ChatPoll/ChatPoll/AccountViewController.swift
/* Copyright (c) 2018 BlackBerry.  All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.

 * This sample code was created by BlackBerry using SDKs from Apple Inc.
 * and may contain code licensed for use only with Apple products.
 * Please review your Apple SDK Agreement for additional details.
 */

import UIKit
import GoogleSignIn
import BBMEnterprise

class AccountViewController: UITableViewController, BBMConnectivityListener, BBMAuthControllerDelegate {
    @IBOutlet weak var serviceStateLabel : UILabel!
    @IBOutlet weak var authTokenStateLabel : UILabel!
    @IBOutlet weak var setupStateLabel : UILabel!
    @IBOutlet weak var regIdLabel : UILabel!
    @IBOutlet weak var domainLabel : UILabel!
    @IBOutlet weak var userEmailLabel : UILabel!
    @IBOutlet weak var serviceConnectivityLabel : UILabel!

    @IBOutlet weak var switchDeviceButton : UIButton!
    @IBOutlet weak var googleSignInButton : GIDSignInButton!
    @IBOutlet weak var signOutButton : UIButton!

    var serviceMonitor : ObservableMonitor!
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)

        //Add ourselves as a BBMConnectivityListener
        BBMEnterpriseService.shared().add(self)

        //The rootController will be used to show the login UI
        ChatPollApp.app().authController().rootController = self

        //Use an observable monitor to monitor service state and auth state changes
        //This only sets up the monitor.  It is activated in viewWillAppear.  Alternatively, you
        //could add this class as a delegate to QuickStartApp.app().authController()
        serviceMonitor = ObservableMonitor(name:"ServiceStateMonitor") {
            [weak self] () -> Void in

            //serviceStarted and authState are observable properties.  This means this block
            //of code will be called each time the monitor activates, and each time either
            //of these property values change
            self?.serviceStateChanged(ChatPollApp.app().authController().serviceStarted)
            self?.authStateChanged(ChatPollApp.app().authController().authState)
       }

        ChatPollApp.app().authController().signInSilently()
    }

    deinit {
        ChatPollApp.app().authController().rootController = nil
        BBMEnterpriseService.shared().remove(self)
    }


    //MARK: View Lifecycle

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        googleSignInButton.isHidden = true
        switchDeviceButton.isEnabled = true
        signOutButton.isHidden = true
        domainLabel.text = SDK_SERVICE_DOMAIN

        //This will activate and run our serviceMonitor which will update all of the UI elements
        //via authStateChanged and serviceStateChanged
        serviceMonitor.activate()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        serviceMonitor.deActivate()
    }


    //MARK: BBMConnectivityListener

    public func connectivityStateChange(_ connected: Bool, strict connectedStrict: Bool) {
        serviceConnectivityLabel.text = connected ? "Connected" : "Disconnected";
    }


    //MARK: BBMAuthControllerDelegate

    func authStateChanged(_ authState: BBMAuthState) {
        authTokenStateLabel.text = authState.authTokenState != nil ? authState.authTokenState! : "No Token"
        setupStateLabel.text = authState.setupState != nil ? authState.setupState! : "Setup Not Started"
        regIdLabel.text = authState.regId != nil ? authState.regId.stringValue : ""

        let email = authState.account != nil ? authState.account.email : ""
        userEmailLabel.text = email

        googleSignInButton.isHidden = ChatPollApp.app().authController().startedAndAuthenticated;
        signOutButton.isHidden = !ChatPollApp.app().authController().startedAndAuthenticated;

        let setupState = authState.setupState != nil ? authState.setupState : ""
        switchDeviceButton.isEnabled = (setupState == kBBMSetupStateDeviceSwitch)

        if (authState.setupState != nil && authState.setupState == kBBMSetupStateFull) {
            ChatPollApp.app().authController().endpointManager.deregisterAnyEndpointAndContinueSetup()
        }

        tableView.reloadData();
    }


    func serviceStateChanged(_ serviceStarted : Bool) {
        serviceStateLabel.text = serviceStarted ? "Started" : "Stopped"
        tableView.reloadData()
    }


    //MARK: IB Actions

    @IBAction func switchDevice(sender: UIButton?) {
        BBMAccess.sendSetupRetry()
    }


    @IBAction func signOut(sender: UIButton?) {
        BBMEnterpriseService.shared().resetService()
        ChatPollApp.app().authController().signOut()
    }

}

